subprojects/core/src/main/java/org/gradle/process/internal/JavaExecHandleBuilder.java

```diff
@@ -311,28 +319,47 @@ protected ArgsEnvironment getEffectiveArgsEnvironment() {
         Map<String, String> environment = getActualEnvironment();
 
         if (hasCommandLineExceedMaxLength(getExecutable(), arguments)) {
-            if (shouldUseEnvironmentVariable(arguments, environment)) {
-                int classPathFlagIndex = arguments.indexOf("-cp");
-                environment.put("CLASSPATH", arguments.get(classPathFlagIndex + 1));
-                arguments.remove(classPathFlagIndex);
-                arguments.remove(classPathFlagIndex);
+            int classPathFlagIndex = arguments.indexOf("-cp");
+            if (shouldTryShorteningCommandLine(arguments)) {
+                if (shouldUseEnvironmentVariable(arguments, environment)) {
+                    environment.put("CLASSPATH", arguments.get(classPathFlagIndex + 1));
+                    arguments.remove(classPathFlagIndex);
+                    arguments.remove(classPathFlagIndex);
+                    LOGGER.info("Gradle is shortening the command line by moving the classpath to the CLASSPATH environment variable.");
+                } else {
+                    try {
+                        File pathingJarFile = File.createTempFile("gradle-javaexec-classpath", ".jar");
+                        List<String> jvmArgs = writePathingJarFile(classpath, pathingJarFile);
```

Comment:
I think instead of parsing out the classpath again from the arguments, we should do some light refactoring:

- We should add a private version of `getAllJvmArgs` that takes a `FileCollection realClasspath`. The existing `getAllJvmArgs` would call this with `this.classpath`.
- We should add a private version of `getAllArguments` that takes a `FileCollection realClasspath`.  This ends up calling the new `getAllJvmArgs(FileCollection)` method.  The existing `getAllArguments` would call this new method with `this.classpath`.
- In `getEffectiveArgsEnvironment`, we'll do what you have here to determine if the command-line is too long.  If we need to shorten it, we'll write the pathing jar and then recall `getAllArguments(realClasspath)` with a _new_ FileCollection that contains just the pathing jar.  That should give us what we want without having to work backwards.

Does that make sense?

Reply:
If we remove the `CLASSPATH` support, then it serve no purpose to have `getEffectiveArgsEnvironment`. We can just use `getEffectiveArguments`. We also don't want to change the normal behaviour of users when calling `getCommandLine`. I will push a change and you tell me.

Reply:
I didn't change much, just reverted to `getEffectiveArguments`.